state-chain/pallets/cf-refunding/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub const PALLET_VERSION: StorageVersion = StorageVersion::new(0);
+
+impl_pallet_safe_mode!(PalletSafeMode; do_refund);
+
+#[frame_support::pallet]
+pub mod pallet {
+	use cf_chains::ForeignChain;
+
+	use super::*;
+	#[pallet::config]
+	#[pallet::disable_frame_system_supertrait_check]
+	pub trait Config: Chainflip {
+		/// Because we want to emit events when there is a config change during
+		/// an runtime upgrade.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Handles egress for all chains.
+		type EgressHandler: EgressApi<AnyChain>;
+
+		/// Safe mode configuration.
+		type SafeMode: Get<PalletSafeMode>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The user does not have enough funds.
+		InsufficientBalance,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RefundScheduled {
+			account_id: ForeignChainAddress,
+			chain: ForeignChain,
+			amount: AssetAmount,
+			epoch: EpochIndex,
+		},
+		RefundIntegrityCheckFailed {
+			epoch: EpochIndex,
+			chain: ForeignChain,
+		},
+	}
+
+	#[pallet::pallet]
+	#[pallet::storage_version(PALLET_VERSION)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Storage for recorded fees per validator and asset.
+	#[pallet::storage]
+	pub type RecordedFees<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		ForeignChain,
+		BTreeMap<ForeignChainAddress, AssetAmount>,
+		OptionQuery,
+	>;
+
+	/// Storage for validator's withheld transaction fees.
+	#[pallet::storage]
+	pub type WithheldTransactionFees<T: Config> =
+		StorageMap<_, Twox64Concat, ForeignChain, AssetAmount, ValueQuery>;
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn record_gas_fee(
+		account_id: ForeignChainAddress,
+		chain: ForeignChain,
+		gas_fee: AssetAmount,
+	) {
+		RecordedFees::<T>::mutate(chain, |maybe_fees| {
+			if let Some(fees) = maybe_fees {
+				fees.entry(account_id).and_modify(|fee| *fee += gas_fee).or_insert(gas_fee);
+			} else {
+				let mut recorded_fees = BTreeMap::new();
+				recorded_fees.insert(account_id, gas_fee);
+				*maybe_fees = Some(recorded_fees);
+			}
+		});
+	}
+	pub fn withheld_transaction_fee(chain: ForeignChain, amount: AssetAmount) {
+		WithheldTransactionFees::<T>::mutate(chain, |fees| *fees += amount);
+	}
+	pub fn on_distribute_withheld_fees(epoch: EpochIndex) {
+		if !T::SafeMode::get().do_refund {
+			log::info!("Refunding is disabled. Skipping refunding.");
+			return;
+		}
+
+		let chains = WithheldTransactionFees::<T>::iter_keys().collect::<Vec<_>>();
+
+		for chain in chains {
+			let mut available_funds = WithheldTransactionFees::<T>::get(chain);
+			// Integrity check before we start refunding
+			if available_funds < RecordedFees::<T>::get(chain).unwrap().values().sum() {
+				log::warn!(
+                    "Integrity check for refunding failed for asset {:?}. Refunding will be skipped.", chain);
```

Comment:
I don't think we should skip in this case: it's possible that we (currently) don't enough funds to fully refund what we owe, but we still want to process refunds if we can.